activerecord/lib/active_record/associations/has_many_association.rb

```diff
@@ -117,8 +115,25 @@ def delete_or_nullify_all_records(method)
 
         # Deletes the records according to the <tt>:dependent</tt> option.
         def delete_records(records, method)
-          if method == :destroy
+          case method
+          when :destroy
             records.each(&:destroy!)
+            update_counter(-records.length) unless reflection.inverse_updates_counter_cache?
+          when :destroy_async
+            association_class = records.first.class
```

Comment:
This is duplicated inside this same class. Let's extract this logic to a private method. Same happens with the other associations.